streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -84,32 +84,35 @@ class ZSink[-R, -InErr, -In, +OutErr, +L, +Z](val channel: ZChannel[R, InErr, Ch
   def collectAllWhileWith[S](z: S)(p: Z => Boolean)(f: (S, Z) => S)(implicit
     ev: L <:< In,
     trace: ZTraceElement
-  ): ZSink[R, InErr, In, OutErr, L, S] =
+  ): ZSink[R, In, E, L, S] =
     new ZSink(
       ZChannel
         .fromZIO(Ref.make(Chunk[In]()).zip(Ref.make(false)))
-        .flatMap { case (leftoversRef, upstreamDoneRef) =>
-          lazy val upstreamMarker: ZChannel[Any, InErr, Chunk[In], Any, InErr, Chunk[In], Any] =
-            ZChannel.readWith(
-              (in: Chunk[In]) => ZChannel.write(in) *> upstreamMarker,
-              ZChannel.fail(_: InErr),
+        .flatMap[R, Nothing, Chunk[In], Any, E, Chunk[L], S] { case (leftoversRef, upstreamDoneRef) =>
```

Comment:
Unfortunately type inference is terrible here with `Nothing` for the `InErr` type. The Scala compiler really does not want to infer `Nothing` here. It is only a problem if you are implementing your own sink operators in terms of channels but it is definitely painful.